contracts/0.8.9/WithdrawalQueueBase.sol

```diff
@@ -0,0 +1,524 @@
+// SPDX-FileCopyrightText: 2023 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+/* See contracts/COMPILERS.md */
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts-v4.4/utils/structs/EnumerableSet.sol";
+import {UnstructuredStorage} from "./lib/UnstructuredStorage.sol";
+import {UnstructuredRefStorage} from "./lib/UnstructuredRefStorage.sol";
+
+/// @title Queue to store and manage WithdrawalRequests.
+/// @dev Use an optimizations to store discounts heavily inspired
+/// by Aragon MiniMe token https://github.com/aragon/aragon-minime/blob/master/contracts/MiniMeToken.sol
+///
+/// @author folkyatina
+abstract contract WithdrawalQueueBase {
+    using EnumerableSet for EnumerableSet.UintSet;
+    using UnstructuredStorage for bytes32;
+
+    /// @notice precision base for share rate and discounting factor values in the contract
+    uint256 public constant E27_PRECISION_BASE = 1e27;
+
+    /// @notice discount factor value that means no discount applying
+    uint96 internal constant NO_DISCOUNT = uint96(E27_PRECISION_BASE);
+
+    // queue for withdrawal requests, indexes (requestId) start from 1
+    bytes32 internal constant QUEUE_POSITION = keccak256("lido.WithdrawalQueue.queue");
+    // length of the queue
+    bytes32 internal constant LAST_REQUEST_ID_POSITION = keccak256("lido.WithdrawalQueue.lastRequestId");
+    // length of the finalized part of the queue. Always <= `requestCounter`
+    bytes32 internal constant LAST_FINALIZED_REQUEST_ID_POSITION =
+        keccak256("lido.WithdrawalQueue.lastFinalizedRequestId");
+    /// finalization discount history, indexes start from 1
+    bytes32 internal constant CHECKPOINTS_POSITION = keccak256("lido.WithdrawalQueue.checkpoints");
+    /// length of the checkpoints
+    bytes32 internal constant LAST_CHECKPOINT_INDEX_POSITION = keccak256("lido.WithdrawalQueue.lastCheckpointIndex");
+    /// amount of eth locked on contract for withdrawal
+    bytes32 internal constant LOCKED_ETHER_AMOUNT_POSITION = keccak256("lido.WithdrawalQueue.lockedEtherAmount");
+    /// withdrawal requests mapped to the owners
+    bytes32 internal constant REQUEST_BY_OWNER_POSITION = keccak256("lido.WithdrawalQueue.requestsByOwner");
+
+    /// @notice structure representing a request for withdrawal.
+    struct WithdrawalRequest {
+        /// @notice sum of the all stETH submitted for withdrawals up to this request
+        uint128 cumulativeStETH;
+        /// @notice sum of the all shares locked for withdrawal up to this request
+        uint128 cumulativeShares;
+        /// @notice address that can claim or transfer the request
+        address payable owner;
+        /// @notice block.timestamp when the request was created
+        uint64 timestamp;
+        /// @notice flag if the request was claimed
+        bool claimed;
```

Comment:
don't need this guy anymore (or maybe for readability)

Reply:
I'd leave it, probably. It's free and more clear.